cachi2/core/package_managers/gomod.py

```diff
@@ -83,7 +95,201 @@ class GoPackage(_GolangModel):
+
+    @property
+    def name(self) -> str:
+        """Get the name for this package based on the parent module's name."""
+        if self.relative_path:
+            return f"{self.module.name}/{self.relative_path}"
+
+        return self.module.name
+
+    @property
+    def real_path(self) -> str:
+        """Get the real path to locate this package on the Internet."""
+        if self.relative_path:
+            return f"{self.module.real_path}/{self.relative_path}"
+
+        return self.module.real_path
+
+    @property
+    def purl(self) -> str:
+        """Get the purl for this package."""
+        return f"pkg:golang/{self.real_path}@{self.module.version}?type=package"
+
+    def to_component(self) -> Component:
+        """Create a SBOM component for this package."""
+        return Component(name=self.name, version=self.module.version, purl=self.purl)
+
+
+class StandardPackage(NamedTuple):
+    """A package from Go standard lib used in the SBOM generation.
+
+    Standard lib packages lack a parent module and, consequentially, a version.
+    """
+
+    name: str
+
+    @property
+    def purl(self) -> str:
+        """Get the purl for this package."""
+        return f"pkg:golang/{self.name}?type=package"
+
+    def to_component(self) -> Component:
+        """Create a SBOM component for this package."""
+        return Component(name=self.name, purl=self.purl)
+
+
+def _create_modules_from_parsed_data(
+    main_module: Module, app_dir: RootedPath, parsed_modules: Iterable[ParsedModule]
+) -> list[Module]:
+    def _create_module(module: ParsedModule) -> Module:
+        if not (replace := module.replace):
+            name = module.path
+            version = module.version or ""
+            original_name = name
+            real_path = name
+        elif replace.version:
+            # module/name v1.0.0 => replace/name v1.2.3
+            name = replace.path
+            version = replace.version
+            original_name = module.path
+            real_path = name
+        else:
+            # module/name v1.0.0 => ./local/path
+            name = module.path
+            resolved_replacement_path = app_dir.join_within_root(module.replace.path)
+            version = _get_golang_version(module.path, resolved_replacement_path)
+            real_path = _resolve_path_for_local_replacement(module)
+            original_name = name
+
+        return Module(name=name, version=version, original_name=original_name, real_path=real_path)
+
+    def _resolve_path_for_local_replacement(module: ParsedModule) -> str:
+        """Resolve all instances of "." and ".." for a local replacement."""
+        if not module.replace:
+            # Should not happen, this function will only be called for replaced modules
+            raise RuntimeError("Can't resolve path for a module that was not replaced")
+
+        path = f"{main_module.real_path}/{app_dir.subpath_from_root}/{module.replace.path}"
+
+        # resolve double dots
```

Comment:
[Pathlib](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve) has a method to automagically do this...

Reply:
Yeah, I just pushed a version that uses it. We were discussing it here: https://github.com/containerbuildsystem/cachi2/pull/140#discussion_r1168819143